source/MaterialXShaderGen/ShaderGenerators/Osl/OslShaderGenerator.cpp

```diff
@@ -117,11 +117,69 @@ ShaderPtr OslShaderGenerator::generate(const string& shaderName, ElementPtr elem
     emitTypeDefs(shader);
     emitFunctionDefinitions(shader);
 
-    emitShaderSignature(shader);
+    // Emit shader type
+    const SgOutputSocket* outputSocket = shader.getNodeGraph()->getOutputSocket();
+    if (outputSocket->type == "surfaceshader")
+    {
+        shader.addStr("surface ");
+    }
+    else if (outputSocket->type == "volumeshader")
+    {
+        shader.addStr("volume ");
+    }
+    else
+    {
+        shader.addStr("shader ");
+    }
+
+    // Emit shader name
+    shader.addStr(shader.getName() + "\n");
+
+    shader.beginScope(Shader::Brackets::PARENTHESES);
+
+    // Emit all app data inputs
+    const Shader::VariableBlock& appDataBlock = shader.getAppDataBlock();
+    for (const Shader::Variable* input : appDataBlock.variableOrder)
+    {
+        const string& type = _syntax->getTypeName(input->type);
+        const string value = _syntax->getTypeDefault(input->type, true);
+        shader.addLine(type + " " + input->name + " = " + value + " [[ int lockgeom=0 ]],");
+    }
+
+    // Emit all uniforms from global scope
+    const Shader::VariableBlock& globalUniformBlock = shader.getUniformBlock(Shader::GLOBAL_SCOPE);
+    for (const Shader::Variable* uniform : globalUniformBlock.variableOrder)
+    {
+        shader.beginLine();
+        emitUniform(*uniform, shader);
+        shader.addStr(",");
+        shader.endLine(false);
+    }
+
+    // Emit all uniforms from shader interface block
```

Comment:
Just curious here for OSL: Are all uniforms just declared unscoped and unstructured ? I guess I'm still a bit confused about shader interface block and global block then :)

Reply:
Yes for OSL all inputs are just declared as a parameter list to the shader function. So here I'm just adding all inputs to the function signature. The only difference for uniforms vs. varyings (appdata) is that varyings should have a meta data tag [[ int lockgeom=0 ]] to indicate it's value is not constant for the whole geometry.

Reply:
Thanks for explaining. I like separation of storage organization from how things are output to the shader.